### PR TITLE
Api Explorer: Styling Refactor

### PR DESCRIPTION
**Pull request by josephaxisa**
I have done some research to ensure that we are doing things in a way that easily allows the introduction of additional themes later on. I came across [this article](https://medium.com/@rossbulat/creating-themes-in-react-with-styled-components-6fce744b4e54) which encouraged me to do a refactor to our styling mostly based on three rules:

1. Styling wrappers should live in one file `ExplorerStyle.tsx`. This keeps component files compact and concise, and easier to read as a result.
2. Any @looker/component overrides that are meant to be defaults should be declared in /src/common/common.tsx and have their name prefixed with `Apix`
3. Not as important but is what I went by here: if it's a couple of styling props then put them inline. If more than a couple, such that prettier forces the component usage to be broken into multiple lines, consider creating a wrapper.  

Open questions: 
- Should CSS template literals be replaced with styled divs? [Here's](https://github.com/looker-open-source/sdk-codegen/blob/b402639a1e0c3446c9224fd6767bcc1a9ecb8774/packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx#L164-L214) an example
- #3 makes me wonder whether styling wrappers should live under the component in the same file or in `ExplorerStyle.tsx`. Would love to hear your opinions. The idea is that once additional themes are introduced, wrappers don't have to be found and validated all over the project.

cc @ssirowy @lukelooker @jkaster @whscullin @miketorosian  @jhardy

### CODE REVIEW

File: packages/api-explorer/src/components/common/common.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+export const ApixTable = styled(Table)``
+ApixTable.defaultProps = { mb: 'large' }
+
+export const ApixCodeBlock = styled(CodeBlock)``
+ApixCodeBlock.defaultProps = {
+  fontSize: 'large',
+  mb: 'larger',
+}
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/common/common.tsx:79]: This is where any @looker/components overrides that are meant to be Apix defaults should be declared. All `Heading` occurrences in this library how now been replaced with `ApixHeading` with local overrides being passed down as props.

It seems that `ApixCode` might not be a good candidate for this file because it doesn't seem to fit wherever we are using `Code` except in `DocMarkdown`. Perhaps we should have a wrapper for `Code` defined in `ExplorerStyle` for use in `DocMarkdown`.

File: packages/api-explorer/src/components/SideNav/SideNavMethods.tsx [hunk 2]
```diff
@@ -44,19 +45,24 @@ export const SideNavMethods: FC<MethodsProps> = ({ methods, tag, specKey }) => {
...
+                  compact
+                  httpMethod={method.httpMethod}
+                >
+                  {method.httpMethod.toUpperCase()}
+                </MethodBadge>
+                <ApixHeading as="h5" mb="0" pt="0" fontWeight="light" truncate>
                   {highlightHTML(pattern, method.summary)}
-                </Heading>
-              </>
+                </ApixHeading>
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/SideNav/SideNavMethods.tsx:64]: Now using the nicer names for endpoint and making use of the `truncate` to trim longer endpoint summaries.

File: packages/api-explorer/src/components/MethodBadge/MethodBadge.tsx [hunk 3]
```diff
@@ -23,59 +23,74 @@
...
+  ${(props) => badgeIntent(getMethodColor(props.httpMethod))};
+  border: 1px solid transparent;
+  border-radius: 4px;
+  font-size: ${({ theme, compact }) =>
+    `${compact ? `calc(${theme.fontSizes.large}/2)` : theme.fontSizes.small}`};
+  font-weight: ${({ theme }) => `${theme.fontWeights.semiBold}`};
+  padding: ${({ theme, compact }) => `${theme.space.xxsmall}
+    ${compact ? '0' : theme.space.xsmall}`};
+  text-align: ${(props) => (props.alignTextCenter ? 'center' : 'left')};
+  min-width: 2.5rem;
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/MethodBadge/MethodBadge.tsx:95]: This component currently breaks `DocMethodSummary` in the landing pages unfortunately. It requires both `httpMethod` to be passed as a prop and as part of `children`. We should refactor it to look more like https://github.com/looker-open-source/sdk-codegen/blob/jax/styling-refactor/packages/api-explorer/src/components/DocMethodSummary/DocSummaryStatus.tsx. At that point, it will just be a `Badge` where the `intent` prop is chosen using `pickIntent` (which also determines the color). 

cc @miketorosian

File: packages/api-explorer/src/components/MainWrapper/MainWrapper.tsx [hunk 4]
```diff
@@ -1,10 +0,0 @@
-import React, { FC } from 'react'
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/MainWrapper/MainWrapper.tsx:1]: This was replaced with https://github.com/looker-open-source/sdk-codegen/pull/254/files#diff-0f5e466453ccc716b22495e3cdac3bcfR119-R121

File: packages/api-explorer/src/components/Header/Header.tsx [hunk 5]
```diff
@@ -26,45 +26,41 @@
...
+import { ApixHeading } from '../common'
+import { Logo } from './svg'
 import { ApiSpecSelector } from './ApiSpecSelector'
 
-const StyledHeaderWrapper = styled(Flex)`
-  border-bottom 1px solid ${(props) => props.theme.colors.palette.charcoal200};
-  flex-direction: row;
-  height: 66px;
-`
-
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/Header/Header.tsx:42]: Moved to `ExplorerStyle`

File: packages/api-explorer/src/components/Header/ApiSpecSelector.tsx [hunk 6]
```diff
@@ -57,11 +52,11 @@ export const ApiSpecSelector: FC<ApiSpecSelectorProps> = ({
...
 
   return (
-    <StyledFieldSelect
+    <FieldSelect
       defaultValue={spec.key}
+      options={options}
       aria-label="Version"
       label="Version"
-      options={options}
       mb="0"
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/Header/ApiSpecSelector.tsx:60]: Wrapper is not needed here.

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 7]
```diff
@@ -1,24 +1,165 @@
...
+      background-color: ${({ theme }) => theme.colors.ui1};
+    }
+  }
+
+  .doc {
+    grid-area: main;
+  }
+`
 
 const reset = css`
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:164]: I am not entirely sure but shouldn't we replace all css with styled divs?

File: packages/api-explorer/src/components/DocTitle/DocTitle.tsx [hunk 8]
```diff
@@ -1,42 +1,32 @@
...
-  title: string
-}
-
-export const DocTitle: FC<DocTitleProps> = ({ title }) => (
-  <Heading
-    as="h1"
-    color="palette.charcoal800"
-    fontSize="xxxlarge"
-    fontWeight="semiBold"
-    mb="xxsmall"
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/DocTitle/DocTitle.tsx:38]: All these props are now `ApixHeading` defaults. There was also a small inconsistency here with `mb` being `xxsmall` instead of `xsmall`

File: packages/api-explorer/src/components/DocStatus/DocStatus.tsx [hunk 9]
```diff
@@ -1,38 +1,80 @@
...
 `
 
 export const DocStatus: FC<DocStatusProps> = ({ method }) => (
-  <StyledBadge className={method.status.toLowerCase()}>
-    {method.status.toUpperCase()}
-  </StyledBadge>
+  <StatusBadgeInternal method={method} className={method.status}>
+    {method.status.toLocaleUpperCase()}
+  </StatusBadgeInternal>
 )
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/DocStatus/DocStatus.tsx:80]: This is very similar to `MethodBadge` and can probably be refactored in a similar fashion. In fact the two are so similar that I think we could combine them.

File: packages/api-explorer/src/components/DocPseudo/DocPseudo.tsx [hunk 10]
```diff
@@ -39,10 +41,12 @@ export const DocPseudo: FC<PseudoProps> = ({ method }) => {
 
   return (
     <Flex alignItems="flex-start">
-      {`${method.operationId}(`}
-      <DocParams parameters={args} />
-      {'): '}
-      <DocPrimaryResponse response={resType} />
+      <Code fontSize="small" fontWeight="normal">
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/DocPseudo/DocPseudo.tsx:44]: `ApixCode` does not fit here. We need to rethink whether we need that to be a global override. cc @miketorosian @jkaster

File: packages/api-explorer/src/ApiExplorer.tsx [hunk 11]
```diff
@@ -112,98 +103,3 @@ const ApiExplorer: FC<ApiExplorerProps> = ({ specs, tryItCallback }) => {
...
-
-    .active & {
-      font-weight: 600;
-    }
-  }
-
-  .doc {
-    grid-area: main;
-  }
-`
```
Comment (josephaxisa) [inline packages/api-explorer/src/ApiExplorer.tsx:209]: All this styling has been moved to `ExplorerStyle.tsx`

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 12]
```diff
@@ -1,24 +1,165 @@
...
+interface SideNavWrapperProps {
+  headerHeight?: string
+}
+
+export const SideNavToggleWrapper = styled.div<SideNavWrapperProps>`
+  position: relative;
+  margin-top: ${parseInt(HEADER_HEIGHT, 10) / 2};
+
+  ${IconButton} {
+    background: #fff;
```
Comment (josephaxisa) [inline packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:60]: @miketorosian Can we leverage the calculated color for the theme here? Otherwise we'll have to introduce additional styling logic once we introduce additional themes.

File: packages/api-explorer/src/components/DocPseudo/DocPseudo.tsx (hunk 10 above)
Comment (jkaster) [reply packages/api-explorer/src/components/DocPseudo/DocPseudo.tsx:44]: I think the "pseudocode" render may have specific styling rules that don't apply elsewhere due to the tooltips, wrapping of parameters, and relatively compact render.

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx (hunk 12 above)
Comment (lukelooker) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:60]:
```suggestion
    background: ${({ theme: { colors }) => colors.background };
```

Should keep you theme-safe :)

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 13]
```diff
@@ -1,24 +1,165 @@
...
+        border: 1px solid ${({ theme }) => theme.colors.ui2};
+      }
+    }
+  }
+
+  ${SidebarItem} {
+    font-size: ${({ theme }) => theme.fontSizes.small};
+
+    & {
+      border-radius: 4px;
```
Comment (lukelooker) [inline packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:147]:
```suggestion
      border-radius: ${({ theme: { radii }}) => radii.medum};;
```

**Comment by jkaster** [comment]
LGTM. It moves us much further along

File: packages/api-explorer/src/scenes/TagScene/TagScene.tsx [hunk 14]
```diff
@@ -52,7 +52,7 @@ export const TagScene: FC<TagSceneProps> = ({ api }) => {
 
   return (
```
Comment (ssirowy) [inline packages/api-explorer/src/scenes/TagScene/TagScene.tsx:53]: Not part of this PR, but I looked at this full file.  On line 45, you filter for the tag, and then take the 0th element.  The more semantic method to use is `find` which will return the your the first object it finds that matches condition.

File: packages/api-explorer/src/components/MethodBadge/MethodBadge.tsx [hunk 15]
```diff
@@ -23,59 +23,74 @@
  SOFTWARE.
 
  */
-
 import React, { FC } from 'react'
-import { Badge } from '@looker/components'
-import styled from 'styled-components'
+import { generatePressed, intentUIBlend } from '@looker/design-tokens'
+import { HttpMethod } from '@looker/sdk/src'
```
[the remark this answers is not in the document]
Comment (jkaster) [reply packages/api-explorer/src/components/MethodBadge/MethodBadge.tsx:28]: That's an errant import location. Should be `@looker/sdk/lib/browser` instead to scope it to browser-only imports

File: packages/api-explorer/src/ApiExplorer.tsx [hunk 16]
```diff
@@ -89,21 +83,18 @@ const ApiExplorer: FC<ApiExplorerProps> = ({ specs, tryItCallback }) => {
...
-          <Box className={isSideNavOpen ? 'doc open' : 'doc'}>
-            <Main>
-              <AppRouter
-                api={spec.api}
-                specKey={spec.key}
-                tryItCallback={tryItCallback}
-              />
-            </Main>
+          </SideNavDivider>
+          <Box className={isSideNavOpen ? 'main doc open' : 'main doc'}>
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/ApiExplorer.tsx:92]: I'd like to get rid of the class names too but I am not sure how to go about doing that. Do you have a reference on how I could do that?

[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/ApiExplorer.tsx:92]: cc @miketorosian since you'll probably do the heavy lifting for us here

File: packages/api-explorer/src/components/DocPseudo/DocParams.tsx [hunk 17]
```diff
@@ -9,12 +33,12 @@ export interface DocArgsProps {
 }
 
 export const DocParams: FC<DocArgsProps> = ({ parameters }) => (
-  <Flex flexWrap="wrap">
+  <>
     {parameters.map((a, index) => (
-      <FlexItem key={index}>
+      <Fragment key={index}>
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/DocPseudo/DocParams.tsx:38]: The shorthand is being used everywhere except for cases like this since it does not support keys

File: packages/api-explorer/src/components/DocSDKs/DocSDKs.tsx [hunk 18]
```diff
@@ -1,31 +1,33 @@
...
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+
  */
 
-import React from 'react'
+import React, { FC } from 'react'
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/DocSDKs/DocSDKs.tsx:27]: I changed this for the sake of consistency because I have been explicitly importing `FC` throughout the whole library and this snuck through. I will swap to the fully scoped version if that's the norm.

File: packages/api-explorer/src/components/DocStatus/DocStatus.tsx [hunk 19]
```diff
@@ -1,38 +1,80 @@
...
   method: IMethod
 }
 
-const StyledBadge = styled(Badge)`
-  border-radius: 4px;
-
-  &.beta {
-    background-color: ${(props) => props.theme.colors.palette.purple100};
-    color: ${(props) => props.theme.colors.palette.purple700};
+const getStatusColor = (status: string) => {
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/DocStatus/DocStatus.tsx:35]: `DocStatus` and `MethodBadge` both need this refactor. It was introduced during a styling round and it's on my todo list. Now is as good a time as any.

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 20]
```diff
@@ -1,24 +1,165 @@
...
+  grid-template-rows: 1fr;
+  grid-template-columns: ${({ open }) =>
+    open ? '20rem 0 50rem' : '1.5rem 0 50rem'};
+  grid-template-areas: 'sidebar divider main';
+  position: relative;
+
+  ${Sidebar} {
+    grid-area: sidebar;
+    width: 20rem;
+    z-index: 0;
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:94]: cc @miketorosian perhaps we don't need this anymore based on the other pending PR?

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 21]
```diff
@@ -1,24 +1,165 @@
...
+    padding: 0 ${({ theme }) => theme.space.medium};
+
+    button {
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:97]: @miketorosian any ideas?

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 22]
```diff
@@ -1,24 +1,165 @@
...
+  .hi {
+    background: yellow;
+  }
+
+  .main {
+    padding: ${({ theme }) => theme.space.xxlarge};
+  }
+
+  ${SidebarGroup} {
+    > div {
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:124]: @miketorosian not sure if we still need this now that we are switching to the accordion

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx [hunk 23]
```diff
@@ -1,24 +1,165 @@
...
+      border-left: 1px dashed ${({ theme }) => theme.colors.ui2};
+      padding: ${({
+        theme: {
+          space: { xxsmall, xsmall },
+        },
+      }) => `${xxsmall} 0 ${xxsmall} ${xsmall}`};
+    }
+
+    .active > span {
```
[the remark this answers is not in the document]
Comment (josephaxisa) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:133]: @miketorosian same as above

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (miketorosian) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:94]: @josephaxisa - This is setting the width of the sidebar. I am not sure what has changed with the addition of accordion. Isn't the accordion just within this sidebar?

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (miketorosian) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:124]: @josephaxisa - We shouldn't. I'll look to remove.

File: packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (miketorosian) [reply packages/api-explorer/src/components/ExplorerStyle/ExplorerStyle.tsx:97]: @josephaxisa - Let me double-check this, I have a suspicion this is old.

File: packages/api-explorer/src/components/SideNav/SideNavMethods.tsx (hunk 2 above)
Comment (miketorosian) [reply packages/api-explorer/src/components/SideNav/SideNavMethods.tsx:64]: @josephaxisa per the other PR we want to look at adding tooltips to only truncated items.

File: packages/api-explorer/src/components/DocStatus/DocStatus.tsx [hunk 24]
```diff
@@ -1,38 +1,80 @@
...
+    default:
+      return 'inform'
   }
+}
 
-  &.stable {
-    background-color: ${(props) => props.theme.colors.palette.green100};
-    color: ${(props) => props.theme.colors.palette.green700};
-  }
+type BadgeIntent =
```
[the remark this answers is not in the document]
Comment (miketorosian) [reply packages/api-explorer/src/components/DocStatus/DocStatus.tsx:50]: Found one typed in Lens.

File: packages/api-explorer/src/components/DocStatus/DocStatus.tsx [hunk 25]
```diff
@@ -1,38 +1,80 @@
...
+  | 'critical'
+  | 'inform'
+  | 'neutral'
+  | 'key'
 
-  &.experimental {
-    background-color: ${(props) => props.theme.colors.palette.yellow000};
-    color: ${(props) => props.theme.colors.palette.yellow600};
-  }
+const statusIntent = (intent: BadgeIntent) =>
```
[the remark this answers is not in the document]
Comment (miketorosian) [reply packages/api-explorer/src/components/DocStatus/DocStatus.tsx:58]: Refactored and removed this.

File: packages/api-explorer/src/components/SideNav/SideNavMethods.tsx (hunk 2 above)
Comment (josephaxisa) [reply packages/api-explorer/src/components/SideNav/SideNavMethods.tsx:64]: I'll address this in another PR.

File: packages/api-explorer/src/scenes/TagScene/TagScene.tsx (hunk 14 above)
Comment (josephaxisa) [reply packages/api-explorer/src/scenes/TagScene/TagScene.tsx:53]: This has been addressed in #255.
